File: 6-Paradigmas/1288-Canhão_de_Destruição-5/1288.c
#include <stdio.h>

int max(int n1, int n2){
    if(n1>n2){
        return n1;
    }
    else{
        return n2;
    }
}

int maxprice(int price[], int weight[], int repeat, int size){
    int maxinum[size+1];
    for (int i = 0; i <= size; i++) {
        maxinum[i] = 0;
    }

    int currp, currw;

    for(int i=0; i<repeat; i++){
        currp=price[i];
        currw=weight[i];

        for(int w=size; w>currw-1; w--){
            maxinum[w]=max(maxinum[w], maxinum[w-currw]+currp);
        }

    }

    return maxinum[size];
}

int main(){
    int tests, repeat, p1, p2, size, castle;
    scanf("%d",&tests);

    for(int r=0; r<tests; r++){
        scanf("%d", &repeat);
        int prices[repeat];
        int weights[repeat];
        for(int i=0; i<repeat; i++){
            scanf("%d", &p1);
            scanf("%d", &p2);
            prices[i]=p1;
            weights[i]=p2;
        }
        scanf("%d", &size);
    scanf("%d", &castle);

    if (maxprice(prices, weights, repeat, size)>=castle){
        printf("Missao completada com sucesso\n");
    }
    else{
        printf("Falha na missao\n");
    }

        
    }
    
    

    return 0;
}

File: 6-Paradigmas/1288.py
def maxprice(price, weight, size, repeat) -> int:

    maxinum=[0 for i in range(0, size+1, +1)]

    for i in range(0, repeat):
        value_p=price[i]
        value_w=weight[i]

        for w in range(size, value_w-1, -1):
            maxinum[w]=max(maxinum[w], maxinum[w-value_w]+value_p)
        
    return maxinum[size]


tests=int(input())
for r in range(0, tests, +1 ):
    repeats=int(input())

    prices=[]
    weights=[]
    for i in range(0,repeats,+1):
        data=input().split(' ')
        prices.append(int(data[0]))
        weights.append(int(data[1]))
    size=int(input())
    castle=int(input())
    if maxprice(prices, weights, size, repeats) >= castle:
        print("Missao completada com sucesso")
    else:
        print("Falha na missao")